New balancer implementation: consistent hash subset

diegows opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
Add Consistent hash balancing to a subset of nodes. It provides the benefits of consistent hashing, but also providing some load balance between some nodes.

diegows commented:
/retest

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -187,6 +187,11 @@ To enable consistent hashing for a backend:
 `nginx.ingress.kubernetes.io/upstream-hash-by`: the nginx variable, text value or any combination thereof to use for consistent hashing. For example `nginx.ingress.kubernetes.io/upstream-hash-by: "$request_uri"` to consistently hash upstream requests by the current request URI.
 
+"subset" hashing can be enabled using 
+`nginx.ingress.kubernetes.io/upstream-hash-type`: subset. This map requests to
```
Is there going to be any other type? If not can we call this something like `nginx.ingress.kubernetes.io/upstream-hash-subset` and make it boolean defaulting to `false`?

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -187,6 +187,11 @@ To enable consistent hashing for a backend:
 
 `nginx.ingress.kubernetes.io/upstream-hash-by`: the nginx variable, text value or any combination thereof to use for consistent hashing. For example `nginx.ingress.kubernetes.io/upstream-hash-by: "$request_uri"` to consistently hash upstream requests by the current request URI.
 
+"subset" hashing can be enabled using 
```
can you add a brief info or link to another pager explaining what subset consistent hashing is?

ElvinEfendi commented:
@diegows any specific reason you're using different packages per annotation? Why do you not use existing `internal/ingress/annotations/upstreamhashby/` to parse the two new annotations you're introducing?

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+  local nodes = {}
+  local nodeset_map = {}
+  local node_sets = {}
+  local subset_size = 3
```
Can you set this default value in controller side when you parse the annotation? That's how it's done in the rest of the code base, so that we can define default values in one place as much as possible instead of scattering them across Lua and Go files.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+
+  -- if we don't have enough nodes, we use reuse nodes in the last set to
+  -- keep the same same on all of them.
+  if #nodes < node_count then
```
This `if` seems unnecessary. If the condition does not hold the `for` below won't do anything anyway.

diegows replied:
you are right :)

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+    return
+  end
+
+  self.peers = backend.endpoints
```
You don't need this, do you?

diegows replied:
fixed.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+local function build_nodeset_map(backend)
+  local nodes = {}
+  local nodeset_map = {}
+  local node_sets = {}
```
Annotation refers to this as `subset` and also the Go code. Can you keep the naming consistent here as well and call this something like `subsets`?

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+  return o
+end
+
+math.randomseed(os.time())
```
We should not do seeding in individual modules, it should be done once only when Nginx worker process starts. Otherwise it can result in weakened randomness.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/chashsubset.lua:
```diff
@@ -0,0 +1,92 @@
+  local nodeset_id = self.instance:find(key)
+  local nodes = self.node_sets[nodeset_id]
+  local node = nodes[math.random(#nodes)]
+  return node
```
I think you can avoid storing endpoint strings as part of `node_sets`. You are storing original endpoints anyway to do the invalidation in `sync` function below. So what you can do is to store the indexes of the endpoints. With relevant changes in `build_nodeset_map` and `new` functions this function then would look something like:

```
local key = util.lua_ngx_var(self.hash_by)
local node = self.instance:find(key)
local subset = self.subsets[node]
local id = subset[math.random(#subset)]
local endpoint = self.endpoints[id]
return endpoint.address .. ":" .. endpoints.port
```

ElvinEfendi commented on rootfs/etc/nginx/lua/util.lua:
```diff
@@ -18,6 +18,25 @@ function _M.get_nodes(endpoints)
+
+  return result
+end
+
```
I don't see these two new functions being used anywhere.

ElvinEfendi commented on rootfs/etc/nginx/lua/test/balancer/chashsubset_test.lua:
```diff
@@ -0,0 +1,35 @@
+    it("uses correct key for given backend", function()
+      _G.ngx = { var = { request_uri = "/alma/armud" }}
+
+      local resty_chash = package.loaded["resty.chash"]
```
this does not seem to be used

diegows commented:
Thanks for your detailed review @ElvinEfendi. Addressed some of the issues, I hope I can finish everything before the end of the week.

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -189,10 +189,20 @@ nginx.ingress.kubernetes.io/auth-realm: "realm string"
 
+There is a special mode of upstream hashing called subset. In this mode,
+upstream servers are grouped into subsets, and stickness works by mapping keys
+to a subset instead of individual upstream servers. It provides a balence between
```
s/balence/balance

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -189,10 +189,20 @@ nginx.ingress.kubernetes.io/auth-realm: "realm string"
 `nginx.ingress.kubernetes.io/upstream-hash-by`: the nginx variable, text value or any combination thereof to use for consistent hashing. For example `nginx.ingress.kubernetes.io/upstream-hash-by: "$request_uri"` to consistently hash upstream requests by the current request URI.
 
+"subset" hashing can be enabled setting 
+`nginx.ingress.kubernetes.io/upstream-hash-subset`: true. This maps requests to
```
please quote true as "true"

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -189,10 +189,20 @@ nginx.ingress.kubernetes.io/auth-realm: "realm string"
+"subset" hashing can be enabled setting 
+`nginx.ingress.kubernetes.io/upstream-hash-subset`: true. This maps requests to
+subset of nodes instead of a single one. `upstream-hash-subset-size` determines
+the size of each subset (default 3).
```
Please also put a link to the example above

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -412,8 +412,8 @@ func buildLoadBalancingConfig(b interface{}, fallbackLoadBalancing string) strin
-	if backend.UpstreamHashBy != "" {
-		return fmt.Sprintf("hash %s consistent;", backend.UpstreamHashBy)
+	if backend.UpstreamHashBy.UpstreamHashBy != "" {
+		return fmt.Sprintf("hash %s consistent;", backend.UpstreamHashBy.UpstreamHashBy)
```
this function is dead, please rebase once https://github.com/kubernetes/ingress-nginx/pull/3478 gets merged

aledbf commented:
@diegows please squash the commits and rebase from master (we fix several e2e flakiness)

diegows commented:
@aledbf @ElvinEfendi Squashed and rebased. I've applied all the feedback you suggested, improved the code, added e2e test. Let me know how it looks now. Thanks!

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -191,10 +191,16 @@ nginx.ingress.kubernetes.io/auth-realm: "realm string"
 
 NGINX supports load balancing by client-server mapping based on [consistent hashing](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash) for a given key. The key can contain text, variables or any combination thereof. This feature allows for request stickiness other than client IP or cookies. The [ketama](http://www.last.fm/user/RJ/journal/2007/04/10/392555/) consistent hashing meth ...
 
+There is a special mode of upstream hashing called subset. In this mode, upstream servers are grouped into subsets, and stickness works by mapping keys to a subset instead of individual upstream servers. It provides a balance between stickness and load distribution.
```
s/stickness/stickiness

ElvinEfendi replied:
It would also be nice to mention that a specific server is chosen uniformly at random from the selected sticky subset.

diegows replied:
Done.

ElvinEfendi commented:
/lgtm
/approve

diegows commented on internal/ingress/controller/controller.go:
```diff
@@ -655,8 +655,11 @@ func (n *NGINXController) createUpstreams(data []*ingress.Ingress, du *ingress.B
+			if upstreams[defBackend].UpstreamHashBy.UpstreamHashBy == "" {
+				upstreams[defBackend].UpstreamHashBy.UpstreamHashBy = anns.UpstreamHashBy.UpstreamHashBy
+				upstreams[defBackend].UpstreamHashBy.UpstreamHashBySubset = anns.UpstreamHashBy.UpstreamHashBySubset
+				upstreams[defBackend].UpstreamHashBy.UpstreamHashBySubsetSize = anns.UpstreamHashBy.UpstreamHashBySubsetSize
```
Because they are different data types.

ElvinEfendi commented:
@diegows please squash your commits, we can then merge this.
